transport/http/agent.go

```diff
@@ -0,0 +1,123 @@
+package http
+
+import (
+	"net/http"
+	"sync"
+
+	"go.uber.org/yarpc/transport"
+	"go.uber.org/yarpc/transport/internal/errors"
+	"go.uber.org/yarpc/transport/peer/hostport"
+)
+
+// Agent keeps track of http peers and the associated client with which the peer will call into.
+type Agent struct {
+	sync.Mutex
+
+	client    *http.Client
+	peerNodes map[string]*peerNode
+}
+
+// peerNode keeps track of a HostPortPeer and any subscribers referencing it
+type peerNode struct {
+	peer       *hostport.Peer
+	references map[transport.PeerSubscriber]struct{}
+}
+
+// NewDefaultAgent creates an http agent with the default parameters
+func NewDefaultAgent() *Agent {
+	return NewAgent(&defaultConfig)
+}
+
+// NewAgent creates a new http agent for managing peers and sending requests
+func NewAgent(cfg *outboundConfig) *Agent {
+	return &Agent{
+		client:    buildClient(cfg),
+		peerNodes: make(map[string]*peerNode),
+	}
+}
+
+// GetClient gets the http client that should be used for making requests
+func (a *Agent) GetClient() *http.Client {
+	return a.client
+}
+
+// RetainPeer gets or creates a Peer for the specificed PeerList
+func (a *Agent) RetainPeer(pid transport.PeerIdentifier, sub transport.PeerSubscriber) (transport.Peer, error) {
```

Comment:
Factoring the PeerSubscriber interface out of PeerList might be obscuring some clarity here. If we don’t need to reuse the subscriber interface, let’s consolidate it.